pymc/gp/hsgp.py

```diff
@@ -0,0 +1,265 @@
+        self.L = L
+        self.m = m
+        self.c = c
+        self.n_dims = cov_func.n_dims
+        self._boundary_set = False
+
+        super().__init__(mean_func=mean_func, cov_func=cov_func)
+
+    def __add__(self, other):
+        raise NotImplementedError("Additive HSGPs aren't supported ")
+
+    def _set_boundary(self, X):
+        """Make L from X and c if L is not passed in."""
+        if not self._boundary_set:
+            if self.L is None:
+                # Define new L based on c and X range
+                La = pt.abs(pt.min(X, axis=0))
+                Lb = pt.abs(pt.max(X, axis=0))
+                self.L = self.c * pt.max(pt.stack((La, Lb)), axis=0)
+            else:
+                self.L = pt.as_tensor_variable(self.L)
+
+            self._boundary_set = True
+
+    def prior_components(self, X: Union[np.ndarray, pt.TensorVariable]):
+        """Return the basis and coefficient priors, whose product is the HSGP.  It can be useful
+        to bypass the GP API and work with the basis and coefficients directly for models with i.e.
+        multiple GPs.  It returns the basis `phi` and the power spectral density, `psd`.  The GP
+        `f` can be formed by `f = phi @ (pm.Normal("hsgp_coeffs", size=psd.size) * psd)`.
+
+        Parameters
+        ----------
+        X: array-like
+            Function input values.
+        """
+        X, _ = self.cov_func._slice(X)
+        self._set_boundary(X)
+
+        omega, phi, m_star = self._eigendecomposition(X, self.L, self.m, self.n_dims)
+        psd = self.cov_func.power_spectral_density(omega)
+
+        i = int(self.drop_first == True)
+        return phi[:, i:], pt.sqrt(psd[i:])
+
+    @staticmethod
+    def _eigendecomposition(X, L, m, n_dims):
+        """Construct the eigenvalues and eigenfunctions of the Laplace operator."""
+        m_star = pt.prod(m)
+        S = np.meshgrid(*[np.arange(1, 1 + m[d]) for d in range(n_dims)])
+        S = np.vstack([s.flatten() for s in S]).T
+        eigvals = pt.square((np.pi * S) / (2 * L))
+        phi = pt.ones((X.shape[0], m_star))
+        for d in range(n_dims):
+            c = 1.0 / np.sqrt(L[d])
+            phi *= c * pt.sin(pt.sqrt(eigvals[:, d]) * (pt.tile(X[:, d][:, None], m_star) + L[d]))
+        omega = pt.sqrt(eigvals)
+        return omega, phi, m_star
+
+    def prior(self, name: str, X: Union[np.ndarray, pt.TensorVariable], *args, **kwargs):
+        R"""
+        Returns the (approximate) GP prior distribution evaluated over the input locations `X`.
+
+        Parameters
+        ----------
+        name: string
+            Name of the random variable
+        X: array-like
+            Function input values.
+        dims: None
+            Dimension name for the GP random variable.
+        """
+
+        phi, sqrt_psd = self.prior_components(X)
+
+        if self.parameterization == "noncentered":
+            self.beta = pm.Normal(f"{name}_hsgp_coeffs_", size=sqrt_psd.size)
```

Comment:
Is it possible to add a possibility for multigp output where size has an additional dimension? So fer I would do this manually via exposed self.prior_components method

Reply:
yep that's what `self.prior_components` is for.  There are a few different ways one could do it.  I think it's better to let the user do it because it would depend upstream on the covariance.

Reply:
That makes sense, and is actually enough for doing arbitrary things